### Imports

In [24]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### read_csv() is used for reading the csv (comma separated values)

In [25]:
df = pd.read_csv(r'C:\Users\prana\PycharmProjects\Data_Science_Files\Datasets\WineDataset.csv')
df

Title  \
0                                    The Guv'nor, Spain   
1     Bread & Butter 'Winemaker's Selection' Chardon...   
2          Oyster Bay Sauvignon Blanc 2022, Marlborough   
3                      Louis Latour Mâcon-Lugny 2021/22   
4     Bread & Butter 'Winemaker's Selection' Pinot N...   
...                                                 ...   
1285    Vouvray Sec 'Expresion de Silex' 2020/21, Loire   
1286        Waimea Estates Pinot Noir Rosé 2021, Nelson   
1287      Wakefield 'Visionary' Cabernet Sauvignon 2010   
1288  Yalumba 'The Menzies' Cabernet Sauvignon 2015/...   
1289     Yalumba 'Virgilius' Viognier 2017, Eden Valley   

                                            Description              Price  \
0     We asked some of our most prized winemakers wo...   £9.99 per bottle   
1     This really does what it says on the tin. It’s...  £15.99 per bottle   
2     Oyster Bay has been an award-winning gold-stan...  £12.49 per bottle   
3     We’ve sold this wine for thirty years – and fo...  £17.99 per bottle   
4     Bread & Butter is that thing that you can coun...  £15.99 per bottle   
...                                                 ...                ...   
1285  Although Denis Meunier is only in his late 20s...  £11.99 per bottle   
1286  Waimea’s vineyards are less than 1km from the ...  £13.99 per bottle   
1287  Made from the very best fruit nourished by 40-...  £85.00 per bottle   
1288  In 1961, Sir Robert Menzies said that Yalumba'...  £37.99 per bottle   
1289  Yalumba's Virgilius is often regarded as Austr...  £34.99 per bottle   

     Capacity               Grape Secondary Grape Varieties       Closure  \
0        75CL         Tempranillo                       NaN  Natural Cork   
1        75CL          Chardonnay                       NaN  Natural Cork   
2        75CL     Sauvignon Blanc                       NaN      Screwcap   
3        75CL          Chardonnay                       NaN  Natural Cork   
4        75CL          Pinot Noir                       NaN  Natural Cork   
...       ...                 ...                       ...           ...   
1285     75CL        Chenin Blanc                       NaN  Natural Cork   
1286     75CL          Pinot Noir                       NaN      Screwcap   
1287     75CL  Cabernet Sauvignon                       NaN      Screwcap   
1288     75CL  Cabernet Sauvignon                       NaN  Natural Cork   
1289     75CL            Viognier                       NaN      Screwcap   

          Country  Unit                                    Characteristics  \
0           Spain  10.5                  Vanilla, Blackberry, Blackcurrant   
1             USA  10.1  Vanilla, Almond, Coconut, Green Apple, Peach, ...   
2     New Zealand   9.8  Tropical Fruit, Gooseberry, Grapefruit, Grass,...   
3          France  10.1                      Peach, Apricot, Floral, Lemon   
4             USA  10.1   Smoke, Black Cherry, Cedar, Raspberry, Red Fruit   
...           ...   ...                                                ...   
1285       France   9.4                   Quince, Green Apple, Lemon, Pear   
1286  New Zealand   9.8          Floral, Raspberry, Red Cherry, Strawberry   
1287    Australia  10.1  Vanilla, Blackberry, Blackcurrant, Cedar, Euca...   
1288    Australia  10.9                  Black Plum, Blackcurrant, Vanilla   
1289    Australia  10.1                Peach, Apricot, Floral, Honeysuckle   

     Per bottle / case / each   Type         ABV           Region  \
0                  per bottle    Red  ABV 14.00%              NaN   
1                  per bottle  White  ABV 13.50%       California   
2                  per bottle  White  ABV 13.00%      Marlborough   
3                  per bottle  White  ABV 13.50%         Burgundy   
4                  per bottle    Red  ABV 13.50%       California   
...                       ...    ...         ...              ...   
1285               per bottle  White  ABV 12.50%           

### head => returns the first 5 rows
### tail => returns the last 5 rows

In [26]:
df.head()

Title  \
0                                 The Guv'nor, Spain   
1  Bread & Butter 'Winemaker's Selection' Chardon...   
2       Oyster Bay Sauvignon Blanc 2022, Marlborough   
3                   Louis Latour Mâcon-Lugny 2021/22   
4  Bread & Butter 'Winemaker's Selection' Pinot N...   

                                         Description              Price  \
0  We asked some of our most prized winemakers wo...   £9.99 per bottle   
1  This really does what it says on the tin. It’s...  £15.99 per bottle   
2  Oyster Bay has been an award-winning gold-stan...  £12.49 per bottle   
3  We’ve sold this wine for thirty years – and fo...  £17.99 per bottle   
4  Bread & Butter is that thing that you can coun...  £15.99 per bottle   

  Capacity            Grape Secondary Grape Varieties       Closure  \
0     75CL      Tempranillo                       NaN  Natural Cork   
1     75CL       Chardonnay                       NaN  Natural Cork   
2     75CL  Sauvignon Blanc                       NaN      Screwcap   
3     75CL       Chardonnay                       NaN  Natural Cork   
4     75CL       Pinot Noir                       NaN  Natural Cork   

       Country  Unit                                    Characteristics  \
0        Spain  10.5                  Vanilla, Blackberry, Blackcurrant   
1          USA  10.1  Vanilla, Almond, Coconut, Green Apple, Peach, ...   
2  New Zealand   9.8  Tropical Fruit, Gooseberry, Grapefruit, Grass,...   
3       France  10.1                      Peach, Apricot, Floral, Lemon   
4          USA  10.1   Smoke, Black Cherry, Cedar, Raspberry, Red Fruit   

  Per bottle / case / each   Type         ABV       Region            Style  \
0               per bottle    Red  ABV 14.00%          NaN     Rich & Juicy   
1               per bottle  White  ABV 13.50%   California    Rich & Toasty   
2               per bottle  White  ABV 13.00%  Marlborough    Crisp & Zesty   
3               per bottle  White  ABV 13.50%     Burgundy   Ripe & Rounded   
4               per bottle    Red  ABV 13.50%   California  Smooth & Mellow   

  Vintage  Appellation  
0      NV          NaN  
1    2021  Napa Valley  
2    2022          NaN  
3    2022        Macon  
4    2021  Napa Valley

In [27]:
df.tail()

Title  \
1285    Vouvray Sec 'Expresion de Silex' 2020/21, Loire   
1286        Waimea Estates Pinot Noir Rosé 2021, Nelson   
1287      Wakefield 'Visionary' Cabernet Sauvignon 2010   
1288  Yalumba 'The Menzies' Cabernet Sauvignon 2015/...   
1289     Yalumba 'Virgilius' Viognier 2017, Eden Valley   

                                            Description              Price  \
1285  Although Denis Meunier is only in his late 20s...  £11.99 per bottle   
1286  Waimea’s vineyards are less than 1km from the ...  £13.99 per bottle   
1287  Made from the very best fruit nourished by 40-...  £85.00 per bottle   
1288  In 1961, Sir Robert Menzies said that Yalumba'...  £37.99 per bottle   
1289  Yalumba's Virgilius is often regarded as Austr...  £34.99 per bottle   

     Capacity               Grape Secondary Grape Varieties       Closure  \
1285     75CL        Chenin Blanc                       NaN  Natural Cork   
1286     75CL          Pinot Noir                       NaN      Screwcap   
1287     75CL  Cabernet Sauvignon                       NaN      Screwcap   
1288     75CL  Cabernet Sauvignon                       NaN  Natural Cork   
1289     75CL            Viognier                       NaN      Screwcap   

          Country  Unit                                    Characteristics  \
1285       France   9.4                   Quince, Green Apple, Lemon, Pear   
1286  New Zealand   9.8          Floral, Raspberry, Red Cherry, Strawberry   
1287    Australia  10.1  Vanilla, Blackberry, Blackcurrant, Cedar, Euca...   
1288    Australia  10.9                  Black Plum, Blackcurrant, Vanilla   
1289    Australia  10.1                Peach, Apricot, Floral, Honeysuckle   

     Per bottle / case / each   Type         ABV           Region  \
1285               per bottle  White  ABV 12.50%            Loire   
1286               per bottle   Rosé  ABV 13.00%           Nelson   
1287               per bottle    Red  ABV 13.50%  South Australia   
1288               per bottle    Red  ABV 14.50%  South Australia   
1289               per bottle  White  ABV 13.50%              NaN   

                      Style Vintage   Appellation  
1285      Aromatic & Floral    2021       Vouvray  
1286         Delicate & Dry    2021           NaN  
1287  Savoury & Full Bodied    2010  Clare Valley  
1288  Savoury & Full Bodied    2016    Coonawarra  
1289      Aromatic & Floral    2017   Eden Valley

### columns => returns the column names

In [28]:
df.columns

Index(['Title', 'Description', 'Price', 'Capacity', 'Grape',
       'Secondary Grape Varieties', 'Closure', 'Country', 'Unit',
       'Characteristics', 'Per bottle / case / each', 'Type', 'ABV', 'Region',
       'Style', 'Vintage', 'Appellation'],
      dtype='object')

### shape => returns the size (m x n)

In [29]:
df.shape

(1290, 17)

### isnull() => checks for null cells
### False => not null
### True => null

In [30]:
df.isnull()

Title  Description  Price  Capacity  Grape  Secondary Grape Varieties  \
0     False        False  False     False  False                       True   
1     False        False  False     False  False                       True   
2     False        False  False     False  False                       True   
3     False        False  False     False  False                       True   
4     False        False  False     False  False                       True   
...     ...          ...    ...       ...    ...                        ...   
1285  False        False  False     False  False                       True   
1286  False        False  False     False  False                       True   
1287  False        False  False     False  False                       True   
1288  False        False  False     False  False                       True   
1289  False        False  False     False  False                       True   

      Closure  Country   Unit  Characteristics  Per bottle / case / each  \
0       False    False  False            False                     False   
1       False    False  False            False                     False   
2       False    False  False            False                     False   
3       False    False  False            False                     False   
4       False    False  False            False                     False   
...       ...      ...    ...              ...                       ...   
1285    False    False  False            False                     False   
1286    False    False  False            False                     False   
1287    False    False  False            False                     False   
1288    False    False  False            False                     False   
1289    False    False  False            False                     False   

       Type    ABV  Region  Style  Vintage  Appellation  
0     False  False    True  False    False         True  
1     False  False   False  False    False        False  
2     False  False   False  False    False         True  
3     False  False   False  False    False        False  
4     False  False   False  False    False        False  
...     ...    ...     ...    ...      ...          ...  
1285  False  False   False  False    False        False  
1286  False  False   False  False    False         True  
1287  False  False   False  False    False        False  
1288  False  False   False  False    False        False  
1289  False  False    True  False    False        False  

[1290 rows x 17 columns]

### isNull().sum() => returns the null cell count 

In [31]:
df.isnull().sum()

Title                          0
Description                    4
Price                          0
Capacity                       0
Grape                         15
Secondary Grape Varieties    802
Closure                       11
Country                        6
Unit                           9
Characteristics               37
Per bottle / case / each       0
Type                           5
ABV                            9
Region                       166
Style                         78
Vintage                        7
Appellation                  646
dtype: int64

### dropna() => drops the row with a null cell (entire row is dropped even if a single cell is null)

In [32]:
df.dropna()

Title  \
13              Miraval Rosé 2021/22, Côtes de Provence   
15    Peyrassol 'Réserve des Templiers' Rosé 2022, C...   
18    Caves d'Esclans 'Whispering Angel' Rosé 2022, ...   
27    Château de Berne ‘Inspiration’ Rosé 2021/22, C...   
28    Domaine La Chautarde Rosé 2022, Côteaux Varois...   
...                                                 ...   
1224  Rutherglen Shelley's Block Marsanne, Viognier,...   
1233                        Sassicaia 2019/20, Bolgheri   
1242  Stag's Leap 'Hands of Time' Red Blend 2018/19,...   
1264                    Torbreck Descendant Shiraz 2012   
1265  Torbreck 'Runrig' Shiraz 2012/2014, Barossa Va...   

                                            Description               Price  \
13    Miraval shot to fame when Angelina Jolie and B...   £22.99 per bottle   
15    La Commanderie de Peyrassol was established in...   £15.99 per bottle   
18    Winemaker Sacha Lachine is a master of the art...   £23.99 per bottle   
27    A stunning property and landmark of Provence, ...   £17.99 per bottle   
28    Although Provence rosé is replicated the world...   £13.99 per bottle   
...                                                 ...                 ...   
1224  Rutherglen has long been known for its fortifi...   £14.99 per bottle   
1233  Sassicaia is widely acknowledged as one of the...  £270.00 per bottle   
1242  Stag’s Leap shocked the world and beat the fin...   £29.99 per bottle   
1264  Torbreck is one of Australia's leading winerie...  £120.00 per bottle   
1265  Torbreck's Runrig Shiraz is blended with a das...  £180.00 per bottle   

     Capacity               Grape  \
13       75CL            Cinsault   
15       75CL            Cinsault   
18       75CL            Grenache   
27       75CL            Grenache   
28       75CL            Grenache   
...       ...                 ...   
1224     75CL            Marsanne   
1233     75CL  Cabernet Sauvignon   
1242     75CL  Cabernet Sauvignon   
1264     75CL              Shiraz   
1265     75CL              Shiraz   

                              Secondary Grape Varieties       Closure  \
13                               Syrah, Grenache, Rolle  Natural Cork   
15                                      Syrah, Grenache  Natural Cork   
18                                      Rolle, Cinsault  Natural Cork   
27                                      Syrah, Cinsault  Natural Cork   
28                            Syrah, Carignan, Cinsault      Screwcap   
...                                                 ...           ...   
1224                                Viognier, Roussanne      Screwcap   
1233                                     Cabernet Franc  Natural Cork   
1242  Petite Sirah, Cabernet Franc, Merlot, Petit Ve...  Natural Cork   
1264                                           Viognier  Natural Cork   
1265                                           Viognier  Natural Cork   

        Country  Unit                                    Characteristics  \
13       France   9.8           Strawberry, Herbaceous, Peach, Raspberry   
15       France   9.8           Strawberry, Herbaceous, Peach, Raspberry   
18       France  10.1           Strawberry, Herbaceous, Peach, Raspberry   
27       France   9.8           Strawberry, Herbaceous, Peach, Raspberry   
28       France   9.8                       Strawberry, Peach, Raspberry   
...         ...   ...                                                ...   
1224  Australia  10.1  Tropical Fruit, Apricot, Butter, Honey, Melon,...   
1233      Italy  10.5                     Spice, Black Fruit, Herbaceous   
1242        USA  11.3  Vanilla, Black Fruit, Black Plum, Blackberry, ...   
1264  Australia  11.3  Violet, Black Cherry, Black Plum, Blackberry, ...   
1265  Australia  11.6  White Pepper, Black Plum, Blackberry, Blackcur...   

     Per bottle / case / each   Type         ABV           Region  \
13                 per bottle   Rosé  ABV 13.00%         Provence   
15                 

### describe() returns the statistical data within a csv

In [33]:
df.describe()

Unit
count  1281.000000
mean     10.177674
std       2.318850
min       0.000000
25%       9.400000
50%      10.100000
75%      10.500000
max      39.000000

### info() => returns the data type and other related information

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1290 entries, 0 to 1289
Data columns (total 17 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Title                      1290 non-null   object 
 1   Description                1286 non-null   object 
 2   Price                      1290 non-null   object 
 3   Capacity                   1290 non-null   object 
 4   Grape                      1275 non-null   object 
 5   Secondary Grape Varieties  488 non-null    object 
 6   Closure                    1279 non-null   object 
 7   Country                    1284 non-null   object 
 8   Unit                       1281 non-null   float64
 9   Characteristics            1253 non-null   object 
 10  Per bottle / case / each   1290 non-null   object 
 11  Type                       1285 non-null   object 
 12  ABV                        1281 non-null   object 
 13  Region                     1124 non-null   objec

### Accessing the columns using [] or .
### eg. df.Unit[0] => returns the first data (10.5 in my dataset)

In [37]:
df.Unit

0       10.5
1       10.1
2        9.8
3       10.1
4       10.1
        ... 
1285     9.4
1286     9.8
1287    10.1
1288    10.9
1289    10.1
Name: Unit, Length: 1290, dtype: float64

In [38]:
df.Unit[0]

np.float64(10.5)

### Deep and Shallow Copy
### view() and copy()

### view() => will keep the copy of the array updated 